How many late days are you using for this assignment? 
4

In [259]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.Requirement already satisfied: pandas in c:\users\raven\appdata\local\packages\pythonsoftwarefoundation.python.3.11_qbz5n2kfra8p0\localcache\local-packages\python311\site-packages (2.2.3)




[notice] A new release of pip is available: 24.0 -> 25.0.1
[notice] To update, run: C:\Users\raven\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [260]:
# Imports and pip installations (if needed)

import pandas as pd

# Part 1: Load the dataset

In [261]:
# Load the given datasets
CKD_cat = pd.read_csv("./chronic_kidney_disease_categorical.csv")
CKD_num = pd.read_csv("./chronic_kidney_disease_numerical.csv")

# Print the data
print(CKD_cat)
print("num")
CKD_num

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

Each entry represents a person. The features are as follows: 
- age 
- blood pressure
- specific gravity
- albumin
- sugar
- red blood cells
- pus cell
- pus cell clumps
- bacteria
- blood glucose random
- blood urea
- serum creatinine
- sodium
- potassium
- hemoglobin
- packed cell volume
- white blood cell count
- red blood cell count
- hypertension
- diabetes mellitus
- coronary artery disease
- appetite
- pedal edema
- anemia

and the labels:

- age
- bp
- sg
- al
- su
- rbc
- pc
- pcc
- ba
- bgr
- bu
- sc
- sod
- pot
- hemo
- pcv
- wbcc
- rbcc
- htn
- dm
- cad
- appet
- pe
- ane

All the features in the numerical dataset are numerical, and most of the features in the categorical dataset are categorical, however some of them are numerical. 

3 numerical examples
- age
- blood pressure
- sodium

3 categorical examples
- albumin
- sugar
- specific gravity

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [ ]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
dup_rows = []
for row in CKD_num.itertuples():
    for row2 in CKD_num.itertuples():
        if row.unique_id == row2.unique_id and row.Index != row2.Index:
            dup_rows.append(row2.Index)

print(len(dup_rows))
# Drop these duplicate rows
CKD_num.drop(dup_rows, inplace=True)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
dup_rows = []
for row in CKD_cat.itertuples():
    for row2 in CKD_cat.itertuples():
        if row.unique_id == row2.unique_id and row.Index != row2.Index:
            dup_rows.append(row2.Index)

print(len(dup_rows))
# Drop these duplicate rows
CKD_cat.drop(dup_rows, inplace=True)
print(CKD_cat)
print(CKD_num)


62
62
     unique_id   al   su       rbc        pc         pcc          ba  htn  dm  \
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no  no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes  no   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no  no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no  no   
5       114717  0.0  0.0    normal    normal  notpresent  notpresent   no  no   
..         ...  ...  ...       ...       ...         ...         ...  ...  ..   
395     397388  2.0  2.0    normal    normal  notpresent     present  yes  no   
396     789906  0.0  0.0    normal    normal  notpresent  notpresent   no  no   
397     490253  0.0  0.0    normal    normal  notpresent  notpresent   no  no   
398     492993  3.0  0.0       NaN    normal  notpresent  notpresent  yes  no   
399     273038  0.0  0.0    normal    normal  notpresent  notpresent   no  no   

     cad appet   pe  

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

I think all the unique ids are present in both dataset because they are two halves of one whole, and when we merge then, we'll get the full picture on each patient

In [263]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
CKD = pd.merge(CKD_num, CKD_cat, on='unique_id', how='inner')

#Print the combined dataset
print(CKD)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  ...   
2       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0  ...   
3       992643  35.0   70.0   82.0   36.0  1.1  150.0  3.5  14.5  52.0  ...   
4       573400  60.0   80.0    NaN    NaN  NaN    NaN  NaN   NaN   NaN  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
356     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0  ...   
357     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
358     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0  ...   
359     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
360     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0  ...   

           pc         pcc          ba  htn   dm cad

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [264]:
# Calculate the percentage of rows that contain atleast one missing value
percent_missing= CKD.isnull().any(axis=1).mean() * 100
# Print %
print(percent_missing)
# Drop these rows from the dataset
CKD.dropna(inplace=True)
# Print the Dataset
print(CKD)

61.772853185595565
     unique_id   age     bp    bgr    bu   sc    sod  pot  hemo   pcv  ...  \
0       203694  38.0   80.0   99.0  19.0  0.5  147.0  3.5  13.6  44.0  ...   
1       421471  37.0   60.0  111.0  35.0  0.8  135.0  4.1  16.2  50.0  ...   
5       109053  64.0   60.0  106.0  27.0  0.7  150.0  3.3  14.4  42.0  ...   
11      349892  59.0   70.0  424.0  55.0  1.7  138.0  4.5  12.6  37.0  ...   
14      630177  58.0   80.0  131.0  18.0  1.1  141.0  3.5  15.8  53.0  ...   
..         ...   ...    ...    ...   ...  ...    ...  ...   ...   ...  ...   
350     486397  20.0   70.0  123.0  44.0  1.0  135.0  3.8  14.6  44.0  ...   
352     462892  36.0   80.0   85.0  16.0  1.1  142.0  4.1  15.6  44.0  ...   
357     475200  64.0  100.0  163.0  54.0  7.2  140.0  4.6   7.9  26.0  ...   
358     168242  48.0   60.0  112.0  44.0  1.2  142.0  4.9  14.5  44.0  ...   
359     474407  21.0   90.0  107.0  40.0  1.7  125.0  3.5   8.3  23.0  ...   

           pc         pcc          ba  htn  

## Part 3.4: Sort the dataset according to the Labels

In [265]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
CKD.sort_values(by="Target", inplace=True, ignore_index=True)

# Print the dataset
print(CKD)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0  90.0  107.0   40.0   1.7  125.0  3.5   8.3  23.0  ...   
1       197452  56.0  70.0  210.0   26.0   1.7  136.0  3.8  16.1  52.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       602404  55.0  90.0  273.0  235.0  14.2  132.0  3.4   8.3  22.0  ...   
4       955830  55.0  80.0  214.0   73.0   3.9  137.0  4.9  10.9  34.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
133     459229  57.0  80.0  133.0   48.0   1.2  147.0  4.3  14.8  46.0  ...   
134     499228  32.0  70.0  100.0   29.0   1.1  142.0  4.5  14.3  43.0  ...   
135     647552  61.0  70.0  120.0   29.0   0.7  137.0  3.5  17.4  52.0  ...   
136     654306  41.0  80.0  112.0   48.0   0.7  140.0  5.0  17.0  52.0  ...   
137     809452  66.0  70.0   94.0   19.0   0.7  135.0  3.9  16.0  41.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [266]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_CKD = pd.get_dummies(CKD, columns=categorical_columns) 

# Print the dataset
print(encoded_CKD)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0  90.0  107.0   40.0   1.7  125.0  3.5   8.3  23.0  ...   
1       197452  56.0  70.0  210.0   26.0   1.7  136.0  3.8  16.1  52.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       602404  55.0  90.0  273.0  235.0  14.2  132.0  3.4   8.3  22.0  ...   
4       955830  55.0  80.0  214.0   73.0   3.9  137.0  4.9  10.9  34.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
133     459229  57.0  80.0  133.0   48.0   1.2  147.0  4.3  14.8  46.0  ...   
134     499228  32.0  70.0  100.0   29.0   1.1  142.0  4.5  14.3  43.0  ...   
135     647552  61.0  70.0  120.0   29.0   0.7  137.0  3.5  17.4  52.0  ...   
136     654306  41.0  80.0  112.0   48.0   0.7  140.0  5.0  17.0  52.0  ...   
137     809452  66.0  70.0   94.0   19.0   0.7  135.0  3.9  16.0  41.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

We would have the use a base n variable, n being the amount of possible values for the categorical variable

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [267]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    if column in CKD.columns:
        mean = CKD[column].mean()
        std = CKD[column].std()
        CKD = CKD[(CKD[column] <= (mean + 3*std)) & (CKD[column] >= (mean - 3*std))]
        
# Print the dataset
print(CKD)

     unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
0       474407  21.0  90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
1       197452  56.0  70.0  210.0   26.0  1.7  136.0  3.8  16.1  52.0  ...   
2       343710  61.0  80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
4       955830  55.0  80.0  214.0   73.0  3.9  137.0  4.9  10.9  34.0  ...   
5       532520  60.0  90.0  105.0   53.0  2.3  136.0  5.2  11.1  33.0  ...   
..         ...   ...   ...    ...    ...  ...    ...  ...   ...   ...  ...   
133     459229  57.0  80.0  133.0   48.0  1.2  147.0  4.3  14.8  46.0  ...   
134     499228  32.0  70.0  100.0   29.0  1.1  142.0  4.5  14.3  43.0  ...   
135     647552  61.0  70.0  120.0   29.0  0.7  137.0  3.5  17.4  52.0  ...   
136     654306  41.0  80.0  112.0   48.0  0.7  140.0  5.0  17.0  52.0  ...   
137     809452  66.0  70.0   94.0   19.0  0.7  135.0  3.9  16.0  41.0  ...   

           pc         pcc          ba  htn   dm  cad appet   pe

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [268]:
# Normalize the all Numerical Attributes in the dataset.

CKD[numerical_columns] = (CKD[numerical_columns] - CKD[numerical_columns].mean())/ CKD[numerical_columns].std()
# Print the dataset

print(CKD)

     unique_id       age        bp       bgr        bu        sc       sod  \
0       474407 -1.658178  1.713005 -0.327027 -0.071135  0.159226 -2.666487   
1       197452  0.506680 -0.302295  1.854327 -0.575382  0.159226 -0.735734   
2       343710  0.815945  0.705355  1.070734  3.818773  1.574573 -0.911257   
4       955830  0.444827  0.705355  1.939040  1.117448  1.574573 -0.560211   
5       532520  0.754092  1.713005 -0.369383  0.397095  0.545230 -0.735734   
..         ...       ...       ...       ...       ...       ...       ...   
133     459229  0.568533  0.705355  0.223606  0.217006 -0.162443  1.195019   
134     499228 -0.977794 -0.302295 -0.475274 -0.467329 -0.226777  0.317404   
135     647552  0.815945 -0.302295 -0.051710 -0.467329 -0.484113 -0.560211   
136     654306 -0.421116  0.705355 -0.221136  0.217006 -0.484113 -0.033642   
137     809452  1.125210 -0.302295 -0.602343 -0.827506 -0.484113 -0.911257   

          pot      hemo       pcv  ...        pc         pcc   

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

The target column should be removed because we're trying to make a model to predict whether or not someone has CKD, we couldn't check the models prediction skills if it already has the answers

In [269]:
#Remove that column
training = CKD.drop('Target', axis=1)
# Print the dataset
print(training)

     unique_id       age        bp       bgr        bu        sc       sod  \
0       474407 -1.658178  1.713005 -0.327027 -0.071135  0.159226 -2.666487   
1       197452  0.506680 -0.302295  1.854327 -0.575382  0.159226 -0.735734   
2       343710  0.815945  0.705355  1.070734  3.818773  1.574573 -0.911257   
4       955830  0.444827  0.705355  1.939040  1.117448  1.574573 -0.560211   
5       532520  0.754092  1.713005 -0.369383  0.397095  0.545230 -0.735734   
..         ...       ...       ...       ...       ...       ...       ...   
133     459229  0.568533  0.705355  0.223606  0.217006 -0.162443  1.195019   
134     499228 -0.977794 -0.302295 -0.475274 -0.467329 -0.226777  0.317404   
135     647552  0.815945 -0.302295 -0.051710 -0.467329 -0.484113 -0.560211   
136     654306 -0.421116  0.705355 -0.221136  0.217006 -0.484113 -0.033642   
137     809452  1.125210 -0.302295 -0.602343 -0.827506 -0.484113 -0.911257   

          pot      hemo       pcv  ...       rbc        pc     

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [270]:
# Export the dataframe to a new csv file
CKD.to_csv('chronic_kidney_disease.csv', index=False)
# Export the dataframe to a new json file
CKD.to_json('chronic_kidney_disease.json')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [271]:
print(CKD.head(15))

    unique_id       age        bp       bgr        bu        sc       sod  \
0      474407 -1.658178  1.713005 -0.327027 -0.071135  0.159226 -2.666487   
1      197452  0.506680 -0.302295  1.854327 -0.575382  0.159226 -0.735734   
2      343710  0.815945  0.705355  1.070734  3.818773  1.574573 -0.911257   
4      955830  0.444827  0.705355  1.939040  1.117448  1.574573 -0.560211   
5      532520  0.754092  1.713005 -0.369383  0.397095  0.545230 -0.735734   
9      481293  0.754092 -1.309945  3.506226 -0.215206  0.159226 -1.788872   
10     546225 -0.111851 -1.309945  0.858952  1.801784  1.188569  0.141881   
14     615697 -2.585974 -1.309945 -0.602343  0.901342 -0.291111 -0.911257   
16     995177  1.310770 -0.302295  1.939040  1.945855  3.118587 -3.544102   
17     923613  1.001504 -1.309945  2.468494  0.577183  1.831908 -0.560211   
19     397388  0.939651  2.720655  3.336800 -0.251223  1.124235  0.492927   
20     843362  1.434476 -1.309945 -0.094066  2.990367  2.475248 -0.735734   

Paste here: 

| unique_id |       age |        bp |       bgr |        bu |       sc |       sod |        pot |      hemo |       pcv | ...        |  pc      |       pcc |      ba  | htn |  dm | cad | appet | pe | ane | Target |
|-----------|-----------|-----------|-----------|-----------|----------|-----------|------------|------------|-----------|------------|----------|-----------|----------|-----|-----|-----|-------|----|-----|--------|
|     474407 | -1.658178 |  1.713005 | -0.327027 | -0.071135 | 0.159226 | -2.666487 |  -1.269790 | -2.525930  | -2.923537 | ...        | abnormal  | present   | present  | no  | no  | no  | good  | no | yes | ckd    |
|     197452 |  0.506680 | -0.302295 |  1.854327 | -0.575382 | 0.159226 | -0.735734 |  -0.806645 |  0.765092  |  1.113728 | ...        | normal    | notpresent| notpresent| no  | no  | no  | good  | no | no  | ckd    |
|     343710 |  0.815945 |  0.705355 |  1.070734 |  3.818773 | 1.574573 | -0.911257 |   1.354700 | -2.779085  | -2.784321 | ...        | abnormal  | notpresent| notpresent| yes | yes | yes | poor  | yes| yes | ckd    |
|     955830 |  0.444827 |  0.705355 |  1.939040 |  1.117448 | 1.574573 | -0.560211 |   0.891555 | -1.428922  | -1.392160 | ...        | abnormal  | present   | present  | yes | yes | no  | good  | yes| no  | ckd    |
|     532520 |  0.754092 |  1.713005 | -0.369383 |  0.397095 | 0.545230 | -0.735734 |   1.354700 | -1.344537  | -1.531377 | ...        | normal    | notpresent| notpresent| no  | no  | no  | good  | no | no  | ckd    |
|     481293 |  0.754092 | -1.309945 |  3.506226 | -0.215206 | 0.159226 | -1.788872 |  -2.041699 | -2.694700  | -2.645105 | ...        | abnormal  | present   | notpresent| yes | no  | no  | poor  | no | yes | ckd    |
|     546225 | -0.111851 | -1.309945 |  0.858952 |  1.801784 | 1.188569 |  0.141881 |  -0.497881 | -1.893041  | -2.227457 | ...        | normal    | notpresent| notpresent| yes | yes | no  | good  | no | no  | ckd    |
|     615697 | -2.585974 | -1.309945 | -0.602343 |  0.901342 | -0.291111| -0.911257 |   0.891555 | -1.850848  | -1.949025 | ...        | abnormal  | notpresent| present  | no  | no  | no  | poor  | no | no  | ckd    |
|     995177 |  1.310770 | -0.302295 |  1.939040 |  1.945855 | 3.118587 | -3.544102 |  -0.652263 | -2.061811  | -2.227457 | ...        | abnormal  | present   | present  | yes | yes | no  | good  | yes| yes | ckd    |
|     923613 |  1.001504 | -1.309945 |  2.468494 |  0.577183 | 1.831908 | -0.560211 |   1.663463 | -2.019619  | -2.088241 | ...        | abnormal  | notpresent| present  | yes | yes | yes | poor  | yes| no  | ckd    |
|     397388 |  0.939651 |  2.720655 |  3.336800 | -0.251223 | 1.124235 |  0.492927 |  -1.269790 | -0.542878  | -0.556864 | ...        | normal    | notpresent| present  | yes | no  | yes | good  | no | no  | ckd    |
|     843362 |  1.434476 | -1.309945 | -0.094066 |  2.990367 | 2.475248 | -0.735734 |   0.891555 | -1.217959  | -1.252944 | ...        | normal    | notpresent| notpresent| yes | yes | no  | poor  | yes| no  | ckd    |
|     828592 |  1.434476 | -0.302295 |  2.044931 |  1.441607 | 1.381571 | -1.262303 |   0.119646 | -1.639885  | -1.531377 | ...        | abnormal  | present   | present  | yes | yes | yes | good  | no | no  | ckd    |
|     438182 |  1.558182 |  0.705355 |  2.764989 |  3.602667 | 2.024910 | -0.384688 |   2.280990 | -1.597693  | -1.531377 | ...        | abnormal  | notpresent| notpresent| yes | yes | yes | good  | no | no  | ckd    |
|     621332 |  0.692239 |  0.705355 |  3.823899 | -0.251223 | -0.098109| -3.193056 |  -1.269790 | -1.639885  | -1.252944 | ...        | normal    | notpresent| notpresent| yes | yes | no  | poor  | no | no  | ckd    |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

If we were putting the data into the LLM programmatically, what we could do is check each row in the converted table to see if that data point can be found in the old table, and if not, find the closest row to it and put it in the converted table

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [272]:
# Code to rename all the columns in the dataset
# Mapping columns
mapping = {
    "age": "Age",
    "bp": "Blood Pressure (mm/Hg)",
    "sg": "Specific Gravity",
    "al": "Albumin",
    "su": "Sugar",
    "rbc": "Red Blood Cells",
    "pc": "Pus Cell",
    "pcc": "Pus Cell Clumps",
    "ba": "Bacteria",
    "bgr": "Blood Glucose Random (mgs/dl)",
    "bu": "Blood Urea (mgs/dl)",
    "sc": "Serum Creatinine (mgs/dl)",
    "sod": "Sodium (mEq/L)",
    "pot": "Potassium (mEq/L)",
    "hemo": "Hemoglobin (gms)",
    "pcv": "Packed Cell Volume",
    "wbcc": "White Blood Cell Count (cells/cmm)",
    "rbcc": "Red Blood Cell Count (millions/cmm)",
    "htn": "Hypertension",
    "dm": "Diabetes Mellitus",
    "cad": "Coronary Artery Disease",
    "appet": "Appetite",
    "pe": "Pedal Edema",
    "ane": "Anemia",
    "class": "CKD or Not CKD"
}

# Renaming columns in the DataFrame
CKD.rename(columns=mapping, inplace=True)

# Display the updated DataFrame
print(CKD.head())


   unique_id       Age  Blood Pressure (mm/Hg)  Blood Glucose Random (mgs/dl)  \
0     474407 -1.658178                1.713005                      -0.327027   
1     197452  0.506680               -0.302295                       1.854327   
2     343710  0.815945                0.705355                       1.070734   
4     955830  0.444827                0.705355                       1.939040   
5     532520  0.754092                1.713005                      -0.369383   

   Blood Urea (mgs/dl)  Serum Creatinine (mgs/dl)  Sodium (mEq/L)  \
0            -0.071135                   0.159226       -2.666487   
1            -0.575382                   0.159226       -0.735734   
2             3.818773                   1.574573       -0.911257   
4             1.117448                   1.574573       -0.560211   
5             0.397095                   0.545230       -0.735734   

   Potassium (mEq/L)  Hemoglobin (gms)  Packed Cell Volume  ...  Pus Cell  \
0          -1.269790 

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [273]:
# Converted SQL to Pandas code
result = CKD.groupby('Target').size().reset_index(name='count')
print(result)

   Target  count
0     ckd     22
1  notckd     98


*Describe what the above code does here*

The code count the amount of entries that are ckd and notckd